src/offset/library.cairo

```diff
@@ -312,307 +277,192 @@ namespace CarbonableOffseter {
 
-        // Get registered owner of the current token index
-        let (high, low) = split_felt(index);
-        let index_uint256 = Uint256(low=low, high=high);
-        let (token_id) = IERC721Enumerable.tokenByIndex(
-            contract_address=contract_address, index=index_uint256
-        );
-        let (owner) = CarbonableOffseter_registered_owner_.read(token_id);
-        let (time) = CarbonableOffseter_registered_time_.read(token_id);
-
-        // Check if user is registered owner
-        let not_eq = is_not_zero(owner - user);
-        if (not_eq == TRUE) {
-            // Stop if index = 0
-            if (index == 0) {
-                return (claimable=0);
-            }
-
-            // Else move on to next index
-            let (add) = _claimable_iter(
-                contract_address=contract_address, user=user, index=index - 1
-            );
-            return (claimable=add);
+        // [Check] total value is not zero
+        let (is_zero) = uint256_eq(total_value, zero);
+        if (is_zero == TRUE) {
+            return (claimable=0);
         }
 
-        // User is the registered owner, then compute claimable for the current token
+        // [Compute] Computed absorption
+        let (time) = CarbonableOffseter_registered_time_.read(address);
         let (initial_absorption) = ICarbonableProject.getAbsorption(
-            contract_address=contract_address, time=time
+            contract_address=contract_address, slot=slot, time=time
         );
         let (final_absorption) = ICarbonableProject.getCurrentAbsorption(
-            contract_address=contract_address
+            contract_address=contract_address, slot=slot
         );
 
-        // [Check] felt overflow
-        let right_order = is_le(initial_absorption, final_absorption);
+        // [Check] Absorption overflow
+        let right_order = is_le_felt(initial_absorption, final_absorption);
         with_attr error_message("CarbonableOffseter: Error while computing claimable") {
             assert right_order = TRUE;
         }
-        let claimable = final_absorption - initial_absorption;
 
-        // Stop if index is null
-        if (index == 0) {
-            return (claimable=claimable);
+        // [Compute] Total absorption, if then return 0
+        let total_absorption_felt = final_absorption - initial_absorption;
+        if (total_absorption_felt == 0) {
+            return (claimable=0);
         }
+        let (total_absorption) = _felt_to_uint(total_absorption_felt);
 
-        // Else move on to next index
-        let (add) = _claimable_iter(contract_address=contract_address, user=user, index=index - 1);
-        return (claimable=claimable + add);
+        // [Compute] Otherwise returns the absorption corresponding to the ratio
+        // Keep quotient_low only since user_value <= total_value
+        let (quotient_low, _, _) = uint256_mul_div_mod(user_value, total_absorption, total_value);
```

Comment:
Would it be useful to define the helper (view) function `total_absorption(user: felt)` ?

Reply:
We didn't have in the 721 version, but it could be a good idea.
I'll implement it

Reply:
Done